### PR TITLE
AP Status

### PR DESCRIPTION
**Pull request by kit-klein**
This PR exposes AP status over a `/ap` endpoint.  It works very similar to the normal `status` endpoint.

```
curl http://localhost:8080/ap
```
Formatted Response
```
{
  "status": "OK",
  "message": "status",
  "payload": {
    "beacon_int": "100",
    "bss": "uap0",
    "bssid": "dc:a6:32:62:4b:0e",
    "cac_time_left_seconds": "N/A",
    "cac_time_seconds": "0",
    "channel": "6",
    "clients": [],
    "dtim_period": "2",
    "freq": "2437",
    "ht_op_mode": "0x0",
    "ieee80211ac": "0",
    "ieee80211ax": "0",
    "ieee80211n": "0",
    "num_sta": "0",
    "num_sta_ht40_intolerant": "0",
    "num_sta_ht_20_mhz": "0",
    "num_sta_ht_no_gf": "0",
    "num_sta_no_ht": "0",
    "num_sta_no_short_preamble": "0",
    "num_sta_no_short_slot_time": "0",
    "num_sta_non_erp": "0",
    "olbc": "0",
    "olbc_ht": "0",
    "phy": "phy0",
    "secondary_channel": "0",
    "ssid": "kinoko-chat",
    "state": "DISABLED"
  }
}
```

AP status works by enabling support for `hostapd_cli` and using its `status` and `list_sta` commands. 

The important fields in the result `payload` are
`state`: ENABLED or DISABLED (we now disable the AP once wifi connects)
`clients`: a list of MAC addresses for connected clients

### CODE REVIEW

File: iotwifi/wpacfg.go [hunk 1]
```diff
@@ -93,6 +93,8 @@ func (wpa *WpaCfg) StartAP() {
 ssid=` + wpa.WpaCfg.HostApdCfg.Ssid + `
 hw_mode=g
 channel=` + wpa.WpaCfg.HostApdCfg.Channel + `
+ctrl_interface=/var/run/hostapd
+ctrl_interface_group=0
```
Comment (kit-klein) [inline iotwifi/wpacfg.go:97]: This enables hostapd_cli

File: iotwifi/wpacfg.go [hunk 2]
```diff
@@ -114,17 +116,49 @@ rsn_pairwise=CCMP`
...
+
+	// get the standard stats
+	stateOut, err := exec.Command("hostapd_cli", "-i", "uap0", "status").Output()
+	if err != nil {
+		wpa.Log.Fatal("Got error checking state: %s", err.Error())
+		return cfgMap, err
+	}
+
+	// Remove the indexing associated with ssid, bssid, and bss
+	stateOut = bytes.ReplaceAll(stateOut, []byte("[0]"), []byte(""))
```
Comment (kit-klein) [inline iotwifi/wpacfg.go:140]: hostapd can support multiple APs so some of the fields include an index.  Since we only use one we can strip them out for clarity.

File: iotwifi/wpacfg.go [hunk 3]
```diff
@@ -114,17 +116,49 @@ rsn_pairwise=CCMP`
...
+	cfgMap = cfgMapper(stateOut)
+
+	// get the list of connected clients
+	clientsOut, err := exec.Command("hostapd_cli", "-i", "uap0", "list_sta").Output()
```
Comment (kit-klein) [inline iotwifi/wpacfg.go:146]: This returns a list of `\n` separated mac address strings representing clients that are connected to the AP.

File: iotwifi/wpacfg.go (hunk 1 above)
Comment (gabe-ochoa) [reply iotwifi/wpacfg.go:97]: Thoughts on adding `WpaCfg.CtrlInterface` and `WpaCfg.CtrlInterfaceGroup` to the config instead of hard coding these values? Should be ok to hardcode them for now for our purposes but if we want to submit this patch upstream it would be good to make it configurable for other devices / distros / configuration

File: iotwifi/iotwifi.go [hunk 4]
```diff
@@ -110,23 +110,29 @@ func RunWifi(log bunyan.Logger, messages chan CmdMessage, cfgLocation string) {
...
 	command.StartDnsmasq()
 
-	// TODO: check to see if we are stuck in a scanning state before
-	// if in a scanning state set a timeout before resetting
+	// monitor for a future connection - shut down AP when it occurs
 	go func() {
 		for {
-			wpacfg.ScanNetworks()
 			time.Sleep(30 * time.Second)
+			if status, ok := wpacfg.Status(); ok == nil && status["wpa_state"] == "COMPLETED" {
```
Comment (gabe-ochoa) [inline iotwifi/iotwifi.go]: Should `wpacfg.Status()` be using the `APStatus()`function that was added below?

Additionally returning `nil` in an `ok` introduces a 3 value boolean. Could we return false instead in the `Status()` implementation?

File: iotwifi/wpacfg.go [hunk 5]
```diff
@@ -57,7 +57,7 @@ func NewWpaCfg(log bunyan.Logger, cfgLocation string) *WpaCfg {
 }
 
 // StartAP starts AP mode.
-func (wpa *WpaCfg) StartAP() {
+func (wpa *WpaCfg) StartAP() (*exec.Cmd) {
```
Comment (gabe-ochoa) [inline iotwifi/wpacfg.go]: Why the change in function signature? I don't see this response used in this PR anywhere.

File: iotwifi/wpacfg.go (hunk 3 above)
Comment (gabe-ochoa) [reply iotwifi/wpacfg.go:146]: There's a good stdlib `strings` package which makes this easier. I'm indifferent here but since we need it in a string form anyway it might be easier to do something like:

```
lines := range strings.Split(string(clientsOut), "\n")
```

File: iotwifi/wpacfg.go (hunk 2 above)
Comment (gabe-ochoa) [reply iotwifi/wpacfg.go:140]: See comment below but since we're doing string operations converting to a string and then using the `strings` package seems appropriate.

File: iotwifi/wpacfg.go [hunk 6]
```diff
@@ -219,8 +253,8 @@ func (wpa *WpaCfg) ConnectNetwork(creds WpaCredentials) (WpaConnection, error) {
 }
 
 // Status returns the WPA wireless status.
-func (wpa *WpaCfg) Status() (map[string]string, error) {
-	cfgMap := make(map[string]string, 0)
+func (wpa *WpaCfg) Status() (map[string]interface{}, error) {
```
Comment (gabe-ochoa) [inline iotwifi/wpacfg.go]: Was this not able to be a string to string map anymore?

**Review by gabe-ochoa — CHANGES_REQUESTED** [review]
Endpoint and functionality looks great. 

A couple changes:
* Using the `strings` package instead of converting from bytes to strings and back again.
* map[string]string -> map[string]interface{}: I'm not seeing the need to move from the concrete type to interface{} in this PR. Was it changed elsewhere?

File: iotwifi/wpacfg.go (hunk 1 above)
Comment (kit-klein) [reply iotwifi/wpacfg.go:97]: I don't think it makes sense as a user configuration since core logic in the service relies on it.

File: iotwifi/wpacfg.go (hunk 5 above)
Comment (kit-klein) [reply iotwifi/wpacfg.go]: Nice catch - this is leftover from an earlier commit that I reverted.

File: iotwifi/iotwifi.go (hunk 4 above)
Comment (kit-klein) [reply iotwifi/iotwifi.go]: .Status is correct in this case since we are looking at the wlan connection (AP is for the AP).  It probably could be modified to .WLANStatus but I didn't want to introduce breaking changes on the endpoints.

File: iotwifi/wpacfg.go (hunk 3 above)
Comment (kit-klein) [reply iotwifi/wpacfg.go:146]: Makes sense.  We need to do the conversion somewhere - seems more reasonable to do it once on the exec.Command.  I was mimicking the pattern below in the cfgMapper func.

File: iotwifi/wpacfg.go (hunk 6 above)
Comment (kit-klein) [reply iotwifi/wpacfg.go]: This function can - another leftover mod from when I planned to embed the AP status in this endpoint.  Will fix

**Comment by kit-klein** [comment]
Nice review - I'll make suggested changes in the am.

File: iotwifi/iotwifi.go (hunk 4 above)
Comment (kit-klein) [reply iotwifi/iotwifi.go]: Good catch on the return - converting ok to err

File: iotwifi/wpacfg.go (hunk 3 above)
Comment (kit-klein) [reply iotwifi/wpacfg.go:146]: As I dig into this more I think it makes sense to let the cfgMapper function do the bytes to string conversion as it's built to handle the raw output from the cli's.  I'll convert the other areas to process as strings asap.

File: iotwifi/iotwifi.go [hunk 7]
```diff
@@ -110,23 +110,29 @@ func RunWifi(log bunyan.Logger, messages chan CmdMessage, cfgLocation string) {
...
 	command.StartDnsmasq()
 
-	// TODO: check to see if we are stuck in a scanning state before
-	// if in a scanning state set a timeout before resetting
+	// monitor for a future connection - shut down AP when it occurs
 	go func() {
 		for {
-			wpacfg.ScanNetworks()
 			time.Sleep(30 * time.Second)
+			if status, err := wpacfg.Status(); err == nil && status["wpa_state"] == "COMPLETED" {
```
Comment (gabe-ochoa) [inline iotwifi/iotwifi.go:126]: Should we log this error when it happens? I don't think we handle a non nil error for the Status() at the moment.

**Review by gabe-ochoa — APPROVED** [review]
One comment about handling an error. Non blocking. 

Lgtm

File: iotwifi/iotwifi.go (hunk 7 above)
Comment (kit-klein) [reply iotwifi/iotwifi.go:126]: I think we're safe to ignore it here.  The transient error won't hurt anything since the next loop will catch it.

This same method is invoked when we hit the /status endpoint.  An error in that case will be logged which I think would give us insight if something was borked.
